trufflesuite/truffle
# Properly handle quotes, backslashes, etc in commands at console

## PR description

Addresses #5838.

This PR makes it so that arguments to Truffle commands at the Truffle console are handled like arguments at the actual command line.  This is intended primarily so that `truffle call` (#5647) will work properly at the Truffle console once it is merged, but it also fixes bugs with existing commands.

For instance, suppose you have a test file `test/meta coin.js`, and you are at the Truffle console, and you want to run specifically the tests in this file and no others.  How would you do that?  Without this PR, there's no straightforward way!  With it, you can just do `test 'test/meta coin.js'` or `test test/meta\ coin.js` or similar.

(This PR has been substantially overhauled, original description is in strikeout below.)

I originally did this using the [shell-quote](https://www.npmjs.com/package/shell-quote) package, but it turned out to have substantial problems (see discussion below), so I ended up rolling my own function to handle shell quotes and escapes.  I put it in `command-utils.js`; it's called `parseQuotesAndEscapes`.  I also wrote a bunch of tests for it.

Quoting and escaping works basically like in bash -- backslash is escape and (outside of a quote) causes whatever is after it to be taken literally; double-quotes cause whatever's inside to be taken literally except for the sequence `\"`, which escapes the double-quote; single-quotes are similar but don't allow any escaping at all.  Other than that, arguments are split on (unescaped, unquoted) whitespace.

However the `parseQuotesAndEscapes` function does allow customizing the escape characters; to avoid problems, on Windows, instead of backslash acting as escape, caret and grave will act as escape.  (We check whether we're on Windows by checking `path.sep`, which might not be the proper way to do it, but makes sense for this case -- it's precisely when backslash is the path separator that we don't want it acting as escape!)

Finally, a word on error handling.  The `parseQuotesAndEscapes` function will error if it finds an unmatched quote, or if the line ends with an escape character.  However, erroring inside `console-child.js` doesn't produce useful results, so instead I added a redundant call to it from `console.js` which exists purely for validation purposes.  Yes, this means we call it twice -- ideally we'd call it once, save the results, and pass those to `console-child.js`, but due to the weird way `console-child.js` starts up, I'm not sure that's possible.  IDK, I didn't really try it; honestly, I'm not sure we really need to worry about that, I expect input strings to be pretty short.

~~Specifically:~~
1. ~~Quotes are now handled.~~
2. ~~Backslash-escaping works.~~
3. ~~Environment variable substitution with `$` works?  That's not actually something I wanted, but the package I used to accomplish this, [shell-quote](https://www.npmjs.com/package/shell-quote), has no good way to turn this off, so whatever, let's go with it.~~

~~(I mean, there are ways to *sort* of turn it off, but no *good* ways that fully work.  Like, I think the closest you could come would be passing `x => "$" + x` for the `env` argument.  However, in that case, while `$STUFF` would remain as `$STUFF` as desired, `${STUFF}` would get transformed into `$STUFF`.  Which is kind of odd and warty.  I figured, why not, instead of dealing with that, let's just allow evironment variable substitution, why not.)~~

~~Anyway, yeah, I just replaced the spot where we split the arguments with spaces with a call to `shellQuote.parse`, and it takes care of the rest.  One note though -- `shellQuote.parse` will return an *object* when it encounters a bash operator so that you can handle those specially; that's very nice, but we *don't* want to handle those specially at Truffle console (because it would be quite a bit of work!), so I just converted those back to their underlying strings before passing them on.~~

~~Btw, there's one more spot in `console.js` where we split the arguments on spaces; however, I opted to leave that alone, as that's just for purposes of identifying the command.  Notionally I could have replaced it with `shellQuote.parse`, but because the use of the code is limited, it seems to me that the existing code works fine.  I did however add a comment noting this fact.~~

## Testing instructions

I added a bunch of tests for `parseQuotesAndEscapes`.  As for manual testing, here's one way, which is pretty much what I actually used:

1. Make a new directory and unbox metacoin there.
5. Rename `test/metacoin.js` to `test/meta coin.js`.
6. Open up `truffle develop`.
7. Type `test test/meta\ coin.js`.  Observe that this works.

By comparison, you can try this on the `develop` branch and observe that you get an error.
